fe/fe-core/src/main/java/org/apache/doris/nereids/rules/join/JoinReorderGreedy.java

```diff
@@ -0,0 +1,136 @@
+    }
+
+    /**
+     * A Group Plan which contains multiple plans.
+     */
+    public class PlanGroupSet {
+        List<Integer> nodes = Lists.newArrayList();
+        BitSet nodesBitSet = new BitSet(32);
+        List<List<Integer>> edges = Lists.newArrayList();
+        double lowestCost = Double.MAX_VALUE;
+        long rowCount = -1;
+
+        public PlanGroupSet(Integer firstInput) {
+            this.nodes.add(firstInput);
+            this.nodesBitSet.set(firstInput);
+        }
+
+        public void addPlan(int newPlan, double lowestCost) {
+            this.edges.add(graph.findEdges(nodesBitSet, newPlan));
+            this.nodes.add(newPlan);
+            this.nodesBitSet.set(newPlan);
+            this.lowestCost = lowestCost;
+        }
+
+        /**
+         * Convert to Plan tree.
+         */
+        public Plan toPlan() {
+            Plan left;
+            left = graph.plan(nodes.get(0));
+            for (int i = 1; i < nodes.size(); i++) {
+                int rightIndex = nodes.get(i);
+                List<Integer> edgeIndex = edges.get(i - 1);
+                left = new LogicalJoin<>(
+                        graph.joinTypes(edgeIndex.get(0)),
+                        graph.conditions(edgeIndex),
+                        left,
+                        graph.plan(rightIndex)
+                );
+            }
+
+            return left;
+        }
+    }
+
+    /**
+     * do greedy.
+     */
+    public PlanGroupSet greedy() {
+        List<PlanGroupSet> result = new ArrayList<>(graph.nodes().size());
+        for (int i = 0; i < graph.nodes().size(); i++) {
+            // i as starting left plannode.
+            // For each starting left plannode, do greedy selection。
+            PlanGroupSet leftPlanGroupSet = new PlanGroupSet(i);
+            result.add(leftPlanGroupSet);
+
+            findLowestCostRight(leftPlanGroupSet);
+        }
+
+        return result.stream()
+                .min(Comparator.comparingDouble(a -> a.lowestCost))
+                .orElseThrow(() -> new RuntimeException("don't exist best plan."));
+    }
+
+    private void findLowestCostRight(PlanGroupSet leftPlans) {
+        if (leftPlans.nodes.size() == graph.nodes().size()) {
+            return;
+        }
+        BitSet rightCandidates = leftPlans.nodes.stream()
+                .map(left -> graph.nodes().get(left).simpleNeighborhood)
+                .reduce(new BitSet(), (lhs, rhs) -> {
+                    lhs.or(rhs);
+                    return lhs;
+                });
+        rightCandidates.andNot(leftPlans.nodesBitSet);
+
+        int right = -1;
+        double lowestCost = Double.MAX_VALUE;
+
+        for (int rightCandidate = rightCandidates.nextSetBit(0); rightCandidate != -1;
+                rightCandidate = rightCandidates.nextSetBit(rightCandidate + 1)) {
+            double cost = getCost(leftPlans, right);
+            if (cost <= lowestCost) {
+                right = rightCandidate;
+            }
```

Comment:
We can stop this recursion if the cost is greater than the lower cost of previous PlanGroupSet